keras/utils/dataset_utils.py

```diff
@@ -14,14 +14,137 @@
 # pylint: disable=g-classes-have-attributes
 
 import multiprocessing
 import os
+import time
 
 import numpy as np
+from tensorflow.python.util.tf_export import keras_export
+
+@keras_export('keras.utils.split_dataset')
+def split_dataset(dataset, 
+                  left_size=None, 
+                  right_size=None, 
+                  shuffle=False, 
+                  seed=None):
+  """Split a dataset into a left half and a right half (e.g. training / validation).
+  
+  Args:
+      dataset: A `tf.data.Dataset` object or 
+               a list/tuple of arrays with the same length.   
+      left_size: If float, it should be in range `[0, 1]` range 
+                 and signifies the fraction of the data to pack in 
+                 the left dataset. If integer, it signifies the number 
+                 of samples to pack in the left dataset. 
+                 If `None`, it defaults to the complement to `right_size`.       
+      right_size: If float, it should be in range `[0, 1]` range 
+                  and signifies the fraction of the data to pack 
+                  in the right dataset. If integer, it signifies 
+                  the number of samples to pack in the right dataset.
+                  If `None`, it defaults to the complement to `left_size`.     
+      shuffle: Boolean, whether to shuffle the data before splitting it.
+      seed: A random seed for shuffling.
+
+  Returns:
+      A tuple of two `tf.data.Dataset` objects: the left and right splits.
+  """
+  
+  if dataset and not isinstance(dataset,(tf.data.Dataset,list)):
+    raise TypeError('`dataset` must be either a tf.data.Dataset object'
+                     'or a list/tuple of arrays',f'Got {type(dataset)}')
+    
+  dataset_as_list = []
+  data_size_warning_flag = False
+  start_time = time.time()
+  
+  for datum in list(dataset):
+    cur_time = time.time()
+    # warns user if the dataset is too large to iterate within 10s
+    if int(cur_time - start_time) > 10 and not data_size_warning_flag:
+      Warning('Takes too long to process the dataset,'
+              'the utility is only meant for small datasets that fit in memory')
+      data_size_warning_flag = True
+    dataset_as_list.append(datum)
+  
+  if shuffle: 
+    if seed:
+      np.random.seed(seed)
+    np.random.shuffle(dataset_as_list)
+    
+  total_size = len(dataset_as_list)
+   
+  # if both left_size and right_size are None, raise error
+  if right_size is None and left_size is None:
+    raise ValueError('Both `left_size` and `right_size`cannot be `None`'
+                     ' either one of them must specified')
+    
+  # if left_size is None, it defaults to the complement to right_size
+  # raises error if right_size not in range [0, 1] or [0, total_size]
+  elif left_size is None and right_size:
+    if type(right_size) == float:
+      if right_size < 0 or right_size > 1:
+        raise ValueError('`right_size` must be in range `[0, 1]`')
```

Comment:
All error messages should specify what the value specified by the user was (for easier debugging)